martin/src/cog/mod.rs

```diff
@@ -0,0 +1,385 @@
+    }
+}
+
+fn rgb_to_png(
+    vec: Vec<u8>,
+    result_dims: (u32, u32),
+    chunk_dims: (u32, u32),
+    chunk_components_count: u32,
+    nodata: Option<u8>,
+    path: &Path,
+) -> Result<Vec<u8>, CogError> {
+    let (data_width, data_height) = chunk_dims;
+    let (tile_width, tile_height) = result_dims;
+    let is_padded = data_width != tile_width;
+    let need_add_alpha = chunk_components_count != 4;
+    let default_value = 0;
+
+    let pixels = if nodata.is_some() || need_add_alpha || is_padded {
+        let mut result_vec = vec![default_value; (tile_width * tile_height * 4) as usize];
+        for row in 0..data_height {
+            'outer: for col in 0..data_width {
+                let idx_chunk =
+                    row * data_width * chunk_components_count + col * chunk_components_count;
+                let idx_result = row * tile_width * 4 + col * 4;
+                for component_idx in 0..chunk_components_count {
+                    if nodata.eq(&Some(vec[(idx_chunk + component_idx) as usize])) {
+                        //This pixel is nodata, just make it transparent and skip it then
+                        let alpha_idx = (idx_result + 3) as usize;
+                        result_vec[alpha_idx] = 0;
+                        continue 'outer;
+                    }
+                    result_vec[(idx_result + component_idx) as usize] =
+                        vec[(idx_chunk + component_idx) as usize];
+                }
+                if need_add_alpha {
+                    let alpha_idx = (idx_result + 3) as usize;
+                    result_vec[alpha_idx] = 255;
+                }
+            }
+        }
+        result_vec
+    } else {
+        vec
+    };
+    let mut result_file_buffer = Vec::new();
+    {
+        let mut encoder = png::Encoder::new(
+            BufWriter::new(&mut result_file_buffer),
+            tile_width,
+            tile_height,
+        );
+        encoder.set_color(png::ColorType::Rgba);
+        encoder.set_depth(png::BitDepth::Eight);
+        let mut writer = encoder
+            .write_header()
+            .map_err(|e| CogError::WritePngHeaderFailed(path.to_path_buf(), e))?;
+        writer
+            .write_image_data(&pixels)
+            .map_err(|e| CogError::WriteToPngFailed(path.to_path_buf(), e))?;
+    }
+    Ok(result_file_buffer)
+}
+
+#[derive(Clone, Debug, Default, PartialEq, Serialize, Deserialize)]
+pub struct CogConfig {
+    #[serde(flatten)]
+    pub unrecognized: UnrecognizedValues,
+}
+
+impl ConfigExtras for CogConfig {
+    fn get_unrecognized(&self) -> &UnrecognizedValues {
+        &self.unrecognized
+    }
+}
+
+impl SourceConfigExtras for CogConfig {
+    async fn new_sources(&self, id: String, path: PathBuf) -> FileResult<Box<dyn Source>> {
+        let tileinfo = TileInfo::new(Format::Png, martin_tile_utils::Encoding::Uncompressed);
+        let meta = get_meta(&path)?;
+        let tilejson = get_tilejson(meta.min_zoom, meta.max_zoom);
```

Comment:
nit: imo, inlining is here more readable. What do you think?

```suggestion
        let tilejson = tilejson! {
            tiles: vec![],
            minzoom: min_zoom,
            maxzoom: max_zoom
        };
```

Reply:
Still working on here. The extent and others would be added to get tile json

Reply:
Thx! 
I forgot to remove the get_tilejson (extent and something else added there, but I decided to add them next PR to make this one not too big).